frontend/dockerfile/parser/errors.go

```diff
@@ -1,37 +1,49 @@
 package parser
 
-import "github.com/pkg/errors"
+import (
+	"github.com/moby/buildkit/util/stack"
+	"github.com/pkg/errors"
+)
 
+// ErrorLocation gives a location in source code that caused the error
 type ErrorLocation struct {
-	Ranges []Range
+	Location []Range
 	error
 }
 
+// Unwrap unwraps to the next error
 func (e *ErrorLocation) Unwrap() error {
 	return e.error
 }
 
+// Range is a code section between two positions
 type Range struct {
 	Start Position
 	End   Position
 }
 
+// Position is a point in source code
 type Position struct {
 	Line      int
 	Character int
 }
```

Comment:
This maps directly to LSP: https://github.com/sourcegraph/go-lsp/blob/4631ffd93a18dc7dd940f937be13589f6ff79826/structures.go#L5-L40

I think at some point we can build a frontend from an IDE and see diagonistics there directly via LSP.

Reply:
Yes, this is where I took it so it would be compatible. Now, who is up to rewriting the Dockerfile parser to LSP?  :)

Reply:
I think a better solution is to write a generic frontend integration for LSP. I think they call it "code actions"? In order to run frontends, diagnose errors, or even a LLB debugger (That's shareable between frontends). You can supplement that with frontend-specific implementations like jump definition, autocomplete, semantic highlighting for Dockerfile.

Reply:
Yes, the errors come from LLB level and using `VertexError` (and probably typed error for every op type in future) here can be mapped to source but to be useful for the user they need to be mapped to original source code.